app/gui/src/controller/searcher.rs

```diff
@@ -374,13 +377,71 @@ impl ThisNode {
 #[derive(Copy, Clone, Debug)]
 #[allow(missing_docs)]
 pub enum Mode {
-    /// Searcher should add a new node at a given position. `source_node` is either a selected node
+    /// Searcher is working with a newly created node. `source_node` is either a selected node
     /// or a node from which the connection was dragged out before being dropped at the scene.
-    NewNode { position: Option<Position>, source_node: Option<ast::Id> },
+    NewNode { node_id: ast::Id, source_node: Option<ast::Id> },
     /// Searcher should edit existing node's expression.
     EditNode { node_id: ast::Id },
 }
 
+impl Mode {
+    /// Return the ID of the node used as target for the Searcher.
+    pub fn node_id(&self) -> ast::Id {
+        match self {
+            Mode::NewNode { node_id, .. } => *node_id,
+            Mode::EditNode { node_id } => *node_id,
+        }
+    }
+
+    fn with_new_node(
+        parameters: SearcherParams,
+        graph: &presenter::Graph,
+        graph_editor: &GraphEditor,
+        graph_controller: &controller::Graph,
+    ) -> FallibleResult<Self> {
```

Comment:
Here is a bit of responsibilities mixing. So far, controllers does not know anything about views  - it was the presenter who know both, set up and call them directly. There was an idea to make controllers, and presenter separate crates (view is already one).